common/CommonAPI.h

```diff
@@ -722,23 +711,281 @@ class CommonAPI
+			creationParams.extent = { w, h, 1 };
+			creationParams.mipLevels = 1;
+			creationParams.arrayLayers = 1;
+			creationParams.usage = imageUsageFlags | nbl::asset::IImage::EUF_TRANSFER_SRC_BIT;
+
+			image = logicalDevice->createImage(std::move(creationParams));
+			auto memReqs = image->getMemoryReqs();
+			memReqs.memoryTypeBits &= logicalDevice->getPhysicalDevice()->getDeviceLocalMemoryTypeBits();
+			logicalDevice->allocate(memReqs, image.get());
+
+			onCreateResourcesWithTripleBufferTarget(image, bufferIx);
+		}
+
+		return image.get();
+	}
+#endif
 public:
 	GraphicalApplication(
 		const std::filesystem::path& _localInputCWD,
 		const std::filesystem::path& _localOutputCWD,
 		const std::filesystem::path& _sharedInputCWD,
 		const std::filesystem::path& _sharedOutputCWD
-	) : nbl::system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
-	void recreateSurface() override
+	) : nbl::system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD),
+		CommonAPI::CommonAPIEventCallback(nullptr, nullptr),
+		m_qRetiredSwapchainResources(),
+		m_imageSwapchainIterations{}
+	{}
+
+	std::unique_lock<std::mutex> recreateSwapchain(
+		uint32_t w, uint32_t h, 
+		nbl::video::ISwapchain::SCreationParams& swapchainCreationParams, 
+		nbl::core::smart_refctd_ptr<nbl::video::ISwapchain>& swapchainRef)
 	{
+		auto logicalDevice = getLogicalDevice();
+		std::unique_lock guard(m_swapchainPtrMutex);
+		CommonAPI::createSwapchain(
+			nbl::core::smart_refctd_ptr<nbl::video::ILogicalDevice>(logicalDevice),
+			swapchainCreationParams, 
+			w, h, 
+			swapchainRef);
+		assert(swapchainRef);
+		m_swapchainIteration++;
+
+		return guard;
 	}
-	void onResize(uint32_t w, uint32_t h) override
+
+	void waitForFrame(
+		uint32_t framesInFlight,
+		nbl::core::smart_refctd_ptr<nbl::video::IGPUFence>& fence)
 	{
+		auto logicalDevice = getLogicalDevice();
+		if (fence)
+		{
+			logicalDevice->blockForFences(1u, &fence.get());
+			if (m_frameIx >= framesInFlight) CommonAPI::dropRetiredSwapchainResources(m_qRetiredSwapchainResources, m_frameIx - framesInFlight);
+		}
+		else
+			fence = logicalDevice->createFence(static_cast<nbl::video::IGPUFence::E_CREATE_FLAGS>(0));
 	}
+
+	nbl::core::smart_refctd_ptr<nbl::video::ISwapchain> acquire(
+		nbl::core::smart_refctd_ptr<nbl::video::ISwapchain>& swapchainRef,
+		nbl::video::IGPUSemaphore* waitSemaphore,
+		uint32_t* imgnum)
+	{
+		nbl::core::smart_refctd_ptr<nbl::video::ISwapchain> swapchain;
+		while (true)
+		{
+			std::unique_lock guard(m_swapchainPtrMutex);
+			swapchain = swapchainRef;
+			if (tryAcquireImage(swapchain.get(), waitSemaphore, imgnum))
+			{
+				return swapchain;
+			}
+		}
+	}
+
+#ifdef ANTI_FLICKER
+	void immediateImagePresent(nbl::video::IGPUQueue* queue, nbl::video::ISwapchain* swapchain, nbl::core::smart_refctd_ptr<nbl::video::IGPUImage>* swapchainImages, uint32_t frameIx, uint32_t lastRenderW, uint32_t lastRenderH)
```

Comment:
could we roll this into `recreateSwapchain` ?

I.e. have an override of `recreateSwapchain` which takes just the extra arguments of `video::IGPUQueue* presentQueue, const uint32_t ` and that would be the one that does the additional triple buffer intermediate blit